keydata.go

```diff
@@ -482,110 +505,172 @@ func (d *KeyData) snapModelAuthKey(auxKey AuxiliaryKey) ([]byte, error) {
-	if data.KDF.Type != kdfType {
-		// Only Argon2i is supported
-		return nil, nil, fmt.Errorf("unexpected KDF type \"%s\"", data.KDF.Type)
-	}
-	if data.Encryption != passphraseEncryption {
+	if d.data.PassphraseParams.Encryption != passphraseEncryption {
 		// Only AES-CFB is supported
-		return nil, nil, fmt.Errorf("unexpected encryption algorithm \"%s\"", data.Encryption)
-	}
-	if data.KeySize > 32 {
-		// The key size can't be larger than 32 with the supported cipher
-		return nil, nil, fmt.Errorf("invalid key size (%d bytes)", data.KeySize)
+		return nil, nil, fmt.Errorf("unexpected encryption algorithm \"%s\"", d.data.PassphraseParams.Encryption)
 	}
 
-	// Derive both the key and IV from the passphrase in a single pass.
-	keyLen := data.KeySize + aes.BlockSize
+	payload = make([]byte, len(d.data.EncryptedPayload))
 
-	params := &KDFCostParams{
-		Time:      uint32(data.KDF.Time),
-		MemoryKiB: uint32(data.KDF.Memory),
-		Threads:   uint8(data.KDF.CPUs)}
-	key, err = kdf.Derive(passphrase, data.KDF.Salt, params, uint32(keyLen))
+	c, err := aes.NewCipher(key)
 	if err != nil {
-		return nil, nil, xerrors.Errorf("cannot derive key from passphrase: %w", err)
-	}
-	if len(key) != keyLen {
-		return nil, nil, errors.New("KDF returned unexpected key length")
+		return nil, nil, xerrors.Errorf("cannot create cipher: %w", err)
 	}
+	stream := cipher.NewCFBDecrypter(c, iv)
+	stream.XORKeyStream(payload, d.data.EncryptedPayload)
 
-	payload = make([]byte, len(data.EncryptedPayload))
+	return payload, authKey, nil
+}
 
-	c, err := aes.NewCipher(key[:data.KeySize])
-	if err != nil {
-		return nil, nil, xerrors.Errorf("cannot create cipher: %w", err)
+func (d *KeyData) platformKeyData() *PlatformKeyData {
+	return &PlatformKeyData{
+		Version:       d.Version(),
+		EncodedHandle: d.data.PlatformHandle,
+		KDFAlg:        crypto.Hash(d.data.KDFAlg),
+		AuthMode:      d.AuthMode(),
 	}
-	stream := cipher.NewCFBDecrypter(c, key[data.KeySize:])
-	stream.XORKeyStream(payload, data.EncryptedPayload)
+}
 
-	return payload, key, nil
+func (d *KeyData) recoverKeysCommon(data []byte) (DiskUnlockKey, PrimaryKey, error) {
+	switch d.Version() {
+	case 0:
+		unlockKey, primaryKey, err := unmarshalV1KeyPayload(data)
+		if err != nil {
+			return nil, nil, &InvalidKeyDataError{xerrors.Errorf("cannot unmarshal cleartext key payload: %w", err)}
+		}
+		return unlockKey, primaryKey, nil
+	case 2:
+		if d.data.KDFAlg != nilHash && !d.data.KDFAlg.Available() {
+			return nil, nil, fmt.Errorf("unavailable KDF digest algorithm %v", d.data.KDFAlg)
+		}
+		pk, err := unmarshalProtectedKeys(data)
+		if err != nil {
+			return nil, nil, &InvalidKeyDataError{xerrors.Errorf("cannot unmarshal cleartext key payload: %w", err)}
+		}
+		return pk.unlockKey(crypto.Hash(d.data.KDFAlg)), pk.Primary, nil
+	default:
+		return nil, nil, fmt.Errorf("invalid keydata version %d", d.Version())
+	}
+}
+
+func (d *KeyData) Version() int {
+	return d.data.Version
```

Comment:
I think having this field as `omitempty` is fine, but I would probably still covert 0 -> 1 here.